File: 121-best-time-to-buy-and-sell-stock/best-time-to-buy-and-sell-stock.java
class Solution {
    public int maxProfit(int[] prices) {
       /* int max = 0;
        for(int i = 0; i < prices.length; i++){
          for(int j = i+1; j < prices.length; j++){
            int profit = prices[j] - prices[i];
            max = Math.max(max,profit);
          }
        }
        return max; */

        int minPrice=Integer.MAX_VALUE;
        int max = 0;
        for(int price : prices){
          if(price < minPrice){
            minPrice = price;
          }else{
            max= Math.max(max,price - minPrice);
          }
        }
        return max;
    }
}

File: 121-best-time-to-buy-and-sell-stock/best-time-to-buy-and-sell-stock.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
         /* BRUTE FORCE 
         int mx = 0;
        for(int i = 0; i < prices.size(); i++){
          for(int j = i+1; j < prices.size(); j++){
            int profit = prices[j] - prices[i];
            mx = max(max,profit);
          }
        }
        return mx; */
        // OPTIMAL SOLUTION
        int minprice=INT_MAX;
        int mx=0;
        for(int price : prices){
          if(price < minprice){
            minprice = price;
          }else{
            mx=max(mx,price - minprice);
          }
        }
        return mx;
    }
};